web/docs/operators/metrics.md

```diff
@@ -109,6 +109,86 @@ Contains a measurement of the amount of memory used by the `tenzir-node` process
 |`swap_space_usage`|`uint64`|The amount of swap space, in bytes. Only available on Linux systems.|
 |`open_fds`|`uint64`|The amount of open file descriptors by the node. Only available on Linux systems.|
 
+### `tenzir.metrics.export`
+
+Contains a measurement of data that goes through the `export` operator, for each data batch.
+
+|Field|Type|Description|
+|:-|:-|:-|
+|`timestamp`|`time`|The time when this event was emitted.|
+|`operator_index`|`uint64`|The index of the `export` operator in the pipeline.|
+|`schema`|`string`|The schema name of the batch.|
+|`schema_id`|`string`|The schema ID of the batch.|
+|`events`|`uint64`|The amount of events that were exported.|
+
+### `tenzir.metrics.import`
+
+Contains a measurement of batch data that goes through the `import` operator, for each data batch.
+
+|Field|Type|Description|
+|:-|:-|:-|
+|`timestamp`|`time`|The time when this event was emitted.|
+|`operator_index`|`uint64`|The index of the `import` operator in the pipeline.|
+|`schema`|`string`|The schema name of the batch.|
+|`schema_id`|`string`|The schema ID of the batch.|
+|`events`|`uint64`|The amount of events that were imported.|
+
+### `tenzir.metrics.publish`
+
+Contains a measurement of data that goes through the `publish` operator, for each data batch.
+
+|Field|Type|Description|
+|:-|:-|:-|
+|`timestamp`|`time`|The time when this event was emitted.|
+|`operator_index`|`uint64`|The index of the `publish` operator in the pipeline.|
+|`topic`|`string`|The topic name.|
+|`schema`|`string`|The schema name of the batch.|
+|`schema_id`|`string`|The schema ID of the batch.|
+|`events`|`uint64`|The amount of events that were published to the `topic`.|
```

Comment:
One thing I just noticed: If values are optional, or negative, it can be easy to infer the wrong type here. Did you consider an approach like this where you seed the series builder with a type explicitly?

```cpp
auto mh = ctrl.metrics_handler(type{"tenzir.metrics.publish", record_type{
  {"topic", string_type{}},
  ...
}});

// ...

mh.emit(record{
  {"topic", topic},
  ...
});

Reply:
I looked into defining something like `metric_handler::emit(const type& t, record&& r) `, but was not satisfied with having to forward a type to `emit()` calls every time.
Your snippet suggests that `ctrl` should create a `metric_handler` for that specific type first, then we can just keep `emit`ting record. I'll look into whether that works.

Reply:
OK, it does work. It just requires some awkward `type` transformations (adding `timestamp`/`operator_id` to all metrics, and `pipeline_id`/`run`/`hidden` for metrics entering the pipeline manager) for the builders to accept the events.

I also aligned the `series_builder` creation mechanism for different metric types in the pipeline manager - the generic metrics do not require a separate `series_builder` in the `analytics_collector` anymore but are part of the operator-specific metric builder map.

Reply:
Great, I think this change is for the better. Having the output schema be inferred can lead to quite a few nasty surprises, I think.